src/ImageHelper.php

```diff
@@ -516,26 +516,63 @@ public static function set_sideload_image_upload_dir(array $upload)
      * The image is expected to be either part of a theme, plugin, or an upload.
      *
      * @param  string $url A URL (absolute or relative) pointing to an image.
-     * @return array       An array (see keys in code below).
+     * @return array<string, mixed> An array (see keys in code below).
      */
-    public static function analyze_url($url)
+    public static function analyze_url(string $url): array
+    {
+        /**
+         * Filters whether to short-circuit the ImageHelper::analyze_url()
+         * file path of a URL located in a theme directory.
+         *
+         * Returning a non-null value from the filter will short-circuit
+         * ImageHelper::analyze_url(), returning that value.
+         *
+         * @since 2.0.0
+         *
+         * @param array<string, mixed>|null $info The URL components array to short-circuit with. Default null.
+         * @param string                    $url  The URL pointing to an image.
+         */
+        $result = \apply_filters('timber/image_helper/pre_analyze_url', null, $url);
+        if (null === $result) {
+            $result = self::_analyze_url($url);
+        }
+
+        /**
+         * Filters the array of anlayzed URL components.
+         *
+         * @since 2.0.0
+         *
+         * @param array<string, mixed> $info The URL components.
+         * @param string               $url  The URL pointing to an image.
+         */
+        return \apply_filters('timber/image_helper/analyze_url', $result, $url);
+    }
+
+    /**
+     * Returns information about a URL.
+     *
+     * @param  string $url A URL (absolute or relative) pointing to an image.
+     * @return array<string, mixed> An array (see keys in code below).
+     */
+    private static function _analyze_url(string $url): array
```

Comment:
According to [PSR-12](https://www.php-fig.org/psr/psr-12/), underscores should not be used for method names.

> Method names MUST NOT be prefixed with a single underscore to indicate protected or private visibility. That is, an underscore prefix explicitly has no meaning.

Yes, we still have some of those in the codebase, but for new private functions, we usually don’t introduce underscore prefixes.

This means that we should find a new name, which is the hard part here. Maybe we can use one of the following?

- `get_url_components()`
- `run_analyze_url()`

Reply:
Renamed `_analyze_url` to `get_url_components`.